proposals/000-settings-api.md

```diff
@@ -0,0 +1,468 @@
+
+#### Methods
+
+* `submitPreferences(preferences: Preferences)`
+    * Submits a new set of `Preferences` to update the current `Settings`.
+    * Note that the `Configurable` might not update its `settings` right away, or might even ignore some of the provided preferences. They are only used as hints to compute the new settings.
+
+#### `Configurable.Settings` Interface
+
+Marker interface for the `Setting` properties holder of a `Configurable`.
+
+### `Setting<V>` Class
+
+Represents a single configurable property of a [Configurable] component and holds its current [value].
+
+:warning: This class **must** be immutable.
+
+For clarity, this reference guide is using a class hierarchy to represent specialization of the `Setting<V>` class, but you could use other solutions. For example in the Kotlin toolkit, there's a unique `Setting<V, E>` class which uses composition to store additional metadata into a `extras: E` property. Specialization are declared as type aliases, e.g. the `RangeSetting` is associated with a `RangeExtras` and declared as `RangeSetting<V> = Setting<V, RangeExtras>`.
+
+#### Properties
+
+* `key: String`
+    * Unique identifier used to serialize [Preferences] to JSON.
+* `value: V`
+    * Current value for this setting.
+* `coder: SettingCoder<V> = IdentitySettingCoder()`
+    * JSON serializer for the [value]
+* `validator: SettingValidator<V> = IdentitySettingValidator()`
+    * Ensures the validity of a [V] value.
+* `activator: SettingActivator = NullSettingActivator()`
+    * Ensures that the condition required for this setting to be active are met in the given [Preferences] – e.g. another setting having a certain preference.
+
+#### Methods
+
+:question: A parameter prefixed with `...` indicates that the function can receive a variable number of arguments.
+
+* `jsonValue() -> Any`
+    * JSON raw representation for the current value.
+    * It is computed using `coder.encode(value)`.
+* `copyFirstValidValueFrom(...candidates: Preferences?, fallback: Setting<V> = this) -> Setting<V>`
+    * Creates a copy of the [Setting] receiver, after replacing its value with the first valid value taken from the given [Preferences] objects, in order.
+    * Each preference is verified using the setting [validator].
+    * If no valid value is found, falls back on the value of `fallback` which defaults to the current value.
+
+### `ToggleSetting` Class (extends `Setting<Boolean>`)
+
+A boolean [Setting].
+
+### `RangeSetting<V>` Class (extends `Setting<V>`)
+
+A [Setting] whose value is constrained to a range.
+
+#### Properties
+
+* `range: Range<V>`
+    * The valid range for the setting value.
+* `suggestedSteps: List<V>?`
+    * Value steps which can be used to decrement or increment the setting. It **must** be sorted in increasing order.
+* `suggestedIncrement: V?`
+    * Suggested value increment which can be used to decrement or increment the setting.
+* `label: (V) -> String`
```

Comment:
I would name this differently. At first glance, I was under the impression that it was a label for the setting. What about `formattedValue` or `prettyValue` ?

Reply:
I agree, since it's a function I would name it `formatValue: (V) -> String`.